### DISCUSSION TITLE
Multi-vendor (marketplace) functionality

### DISCUSSION THREAD

**Discussion by lautsevich** [lovata/oc-shopaholic-plugin#350] [Ideas]
Hi everyone!

We start gathering requirements for the implementation in Shopaholic multi-vendor functionality. Don't hesitate to provide your thoughts and experience it this domain.

P.S. Please, vote for this feature as well:
* 👍 — I need this in a current project.
* ❤️ — I'll need this in the future project.
* 👀 — I needed this in a recent project.

**Comment by alvaro-canepa** [comment]
I actually has an extension plugin to cover some vendor aspects. But it is made for a custom project, so, I can't publish it in the Marketplace. 
I use Backend users as Vendors, selecting a custom group.
Them, extends Order model to add `vendor` relation, using `manager_id` key. 

```php
use Backend\Models\User;
class OrderModel extends ModelBehavior
{
    /**
     * OrderModel constructor.
     *
     * @param \Lovata\OrdersShopaholic\Models\Order $model
     */
    public function __construct($model)
    {
        parent::__construct($model);
        $model->belongsTo['vendor'] = [
            User::class,
            'key' => 'manager_id'
        ];
        $model->bindEvent(
            'model.beforeSave',
            function () use ($model) {
                $this->generateOrderNumberExt();
            }
        );
    }
}
```

`Product` model has been extended too, to has vendors be own of your products .
```php
class ProductModel extends ModelBehavior
{
    /**
     * ProductModel constructor.
     *
     * @param \Lovata\Shopaholic\Models\Product $model
     *
     * @throws \ApplicationException
     */
    public function __construct($model)
    {
        parent::__construct($model);

        $model->belongsToMany['users'] = [
            User::class,
            'table' => 'planetadeleste_vendorsshopaholic_users_products',
            'scope' => 'isNotSuperUser'
        ];
    }
}
```

And a lot of more events and controllers extensions.

**Comment by it-werxdotca** [comment]
I would be very interested in this feature. Understanding that I would have to work with the permissions to get it to work. Maybe a list of things that have interaction with the permissions would be helpful?

**Comment by lautsevich** [comment]
@it-werxdotca, hi! Yep, could you please share your thoughts about the possible functionality of this plugin?

**Comment by mkleanthous** [comment]
I am also very interested in this extension. Some helpful features will be:
- Sale stats for each vendor per product/category/month/year
- Different percentage commission for each vendor/group
- Vendor payment requests to transfer their earnings. Approved by marketplace owner and automatically transferred to bank/paypal/other account
- Separated media folder for each vendor to manage his eshop media content
- If possible, maybe at later stage, ability for design customisations/templating for each vendor to give a different look for his shop

**Comment by buritai** [comment]
I am also very interested in this extension. I have used shopaholic for many projects and currently I have to face a Multivendor ecommerce. Do you have plans of a realease of the extension? It would be the ideal time to integrate this new shopaholic feature in my actual project roadmap.

**Comment by lautsevich** [comment]
Hi @elviofernandez, we don't have a specific release date for this. But AFAIK, one of the October CMS vendor's develops such a plugin since last Summer. In case it will have been released, we'll inform you about it in this issue.

**Comment by buritai** [comment]
Hi **@lautsevich**, Do you know what is the state of this extension? If you can contact this vendor you can tell them that we are willing to pay for the plugin. Our intention is not to start from scratch. The company I work, has built its entire ecommerce line with Shopaholic. It would be very interesting to achieve an integration of multivendor for Shopaholic. Another option to keep in mind is that if extension is not in release stage or is not complete, we can contribute to development and testing.

**Comment by buzmate** [comment]
Hello elviofernandez. As @lautsevich said, we have been working on this since last summer. We don't wont to get paid, we prefer to collaborate on the development of this plugin to speed it's production as much as possible. Shall we setup a team and work together on this?

**Comment by buritai** [comment]
Hi **@buzmate**, please send me an email to elvio.fernandez@squad.com.ar so we can continue privately

**Comment by itwerxdotca** [comment]
@buzmate Will this be in a private repository or public?

**Comment by itwerxdotca** [comment]
There is a Laravel project that runs on 5.6. Though the "multi-vendor" part is sold as an add-on to the core.
https://aimeos.org

I found it to be ver clumsy to work with and documentation scattered. But the code is there for possible hints on how to make things work?

**Comment by itwerxdotca** [comment]
Another idea... Something like the Blog Front plugin? That seems like less hassle then trying to reinvent the horse?

```
Features

Author settings per user

Restrict what categories user can post to
Allow publishing or only draft post creation ( backend user will then need to publish post )
Set back-end user that user is authoring as.
Option to restrict author editing to only posts they create.
Option NOT to use "Author Settings" allows you to customize and protect forms as you see fit
Email post notifications

Send to one or multiple user groups
Included a link back to post page
Sends both HTML and plain text email
HTML email will contain post content
Set groups that can receive notices in component settings
```

**Comment by buzmate** [comment]
> @buzmate Will this be in a private repository or public?

This will be available to public in some way, either through OC marketplace or  a public repo. But still needs a good amount of work to reach the beta version.
